### PR TITLE
feat: add allow list to API keys

### LINKED ISSUE #19854: SDK (Go): codersdk types for scopes[] + allow_list[]
Description

- Update `codersdk.APIKey` and related request/response structs to use `[]APIKeyScope` (or `[]string`) and `[]string allow_list`.
- Preserve `application_connect` and `all` for backward compatibility; mark legacy single `scope` as deprecated in code comments.

Key files/areas

- `codersdk/apikey.go` and dependents.
- Regenerate TS types for the site via existing pipeline.

Acceptance criteria

- `site/src/api/typesGenerated.ts` reflects arrays for `scopes` and `allow_list`.
- CLI and UI compile against the new SDK without regressions.

### PR DESCRIPTION
**Pull request by ThomasK33**
Add API key allow list to the SDK

This PR adds an allow list to API keys in the SDK. The allow list is a list of targets that the API key is allowed to access. If the allow list is empty, a default allow list with a single entry that allows access to all resources is created.

The changes include:

- Adding a default allow list when generating an API key if none is provided
- Adding allow list to the API key response in the SDK
- Converting database allow list entries to SDK format in the API response
- Adding tests to verify the default allow list behavior

Fixes #19854

### CODE REVIEW

**Comment by ThomasK33** [comment]
* **#19972** <a href="https://app.graphite.dev/github/pr/coder/coder/19972?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 3  dependent PRs ([#19983](https://github.com/coder/coder/pull/19983) <a href="https://app.graphite.dev/github/pr/coder/coder/19983?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#19985](https://github.com/coder/coder/pull/19985) <a href="https://app.graphite.dev/github/pr/coder/coder/19985?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#20444](https://github.com/coder/coder/pull/20444) <a href="https://app.graphite.dev/github/pr/coder/coder/20444?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈 <a href="https://app.graphite.dev/github/pr/coder/coder/19972?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#19964** <a href="https://app.graphite.dev/github/pr/coder/coder/19964?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20032** <a href="https://app.graphite.dev/github/pr/coder/coder/20032?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: coderd/apikey/apikey.go [hunk 1]
```diff
@@ -97,6 +97,10 @@ func Generate(params CreateParams) (database.InsertAPIKeyParams, string, error)
 		}
 	}
 
+	if len(params.AllowList) == 0 {
+		params.AllowList = database.AllowList{database.AllowListTarget{}}
```
Comment (Emyrk) [inline coderd/apikey/apikey.go]: This is where should be explicit. Not assuming an empty target is "Everything". This code syntactically looks like the target does nothing.

```golang
	if len(params.AllowList) == 0 {
		params.AllowList = database.AllowList{database.AllowListTarget{type:"*", ID:"*"}}
	}
```

File: coderd/users.go [hunk 2]
```diff
@@ -1588,6 +1589,18 @@ func convertAPIKey(k database.APIKey) codersdk.APIKey {
...
+	allowList := make([]codersdk.APIAllowListTarget, 0, len(k.AllowList))
+	for _, entry := range k.AllowList {
+		var target codersdk.APIAllowListTarget
+		if resource, ok := entry.Type.Value(); ok {
+			target.Type = wildcard.Of(codersdk.RBACResource(resource))
+		}
+		if id, ok := entry.ID.Value(); ok {
+			target.ID = wildcard.Of(id)
+		}
+		allowList = append(allowList, target)
```
Comment (Emyrk) [inline coderd/users.go]: You can write a converter in the `db2sdk` package.

Then do something like [this](https://github.com/coder/coder/blob/c50a0ac749cf99926d14ea8c4eae50f0432c95c2/coderd/users.go#L1372-L1372)

```golang
db2sdk.List(k.AllowList, db2sdk.APIAllowListTarget))
```

Comment (ThomasK33) [reply coderd/users.go]: done

File: coderd/database/db2sdk/db2sdk.go [hunk 3]
```diff
@@ -51,6 +51,17 @@ func ListLazy[F any, T any](convert func(F) T) func(list []F) []T {
...
+func APIAllowListTarget(entry rbac.AllowListElement) codersdk.APIAllowListTarget {
+	target := codersdk.AllowAllTarget()
+	if entry.Type != "" {
+		target.Type = codersdk.RBACResource(entry.Type)
+	}
+	if entry.ID != "" {
+		target.ID = entry.ID
+	}
+	return target
+}
```
Comment (Emyrk) [inline coderd/database/db2sdk/db2sdk.go:59]: I know we have to figure out the semantics of an empty string for the `id`, but it should not be treated as a wildcard imo.

I think we should just pass the rbac entry to the SDK as is. Do not do: `target := codersdk.AllowAllTarget()` as the default.

File: coderd/users.go [hunk 4]
```diff
@@ -1596,6 +1596,11 @@ func convertAPIKey(k database.APIKey) codersdk.APIKey {
 		scopes = append(scopes, codersdk.APIKeyScope(s))
 	}
 
+	allowList := db2sdk.List(k.AllowList, db2sdk.APIAllowListTarget)
+	if len(allowList) == 0 {
+		allowList = append(allowList, codersdk.AllowAllTarget())
```
Comment (Emyrk) [inline coderd/users.go]: I still think we should never have an empty allow list. The db migration handles this iirc, so defaulting to `all` if the list is empty feels like it's hiding some other issue.

Comment (ThomasK33) [reply coderd/users.go]: Yeah, that's true. I've added an assert here since something else must have gone horribly wrong then.

File: coderd/database/db2sdk/db2sdk.go (hunk 3 above)
Comment (ThomasK33) [reply coderd/database/db2sdk/db2sdk.go:59]: updated

File: coderd/users.go [hunk 5]
```diff
@@ -1596,6 +1596,11 @@ func convertAPIKey(k database.APIKey) codersdk.APIKey {
 		scopes = append(scopes, codersdk.APIKeyScope(s))
 	}
 
+	allowList := db2sdk.List(k.AllowList, db2sdk.APIAllowListTarget)
+	if len(allowList) == 0 {
+		panic(fmt.Sprintf("developer error: API key %s has empty allow list", k.ID))
+	}
```
Comment (Emyrk) [inline coderd/users.go]: I know it is annoying, but we should never panic. Should either return an error, or just leave the list empty

**Review by Emyrk — CHANGES_REQUESTED** [review]

File: coderd/users.go (hunk 5 above)
Comment (ThomasK33) [reply coderd/users.go]: Why not?

Comment (Emyrk) [reply coderd/users.go]: The caller of this function is always an http route, which is able to handle and properly return an error.

This should be an internal server error. I know we have a recover MW, but that is a last resort to catch uncaught errors.

Comment (ThomasK33) [reply coderd/users.go]: Right, but this isn't just about returning or bubbling up an error—it's about making our assumptions explicit. We assume that an empty allow list will never be stored in the database.

Therefore, this code path should only occur if something is seriously wrong with our database. In such a case, I don't think we should treat it as a typical application error. Instead, we should signal clearly that our foundational assumptions have been violated and something beyond our control is wrong.

Comment (Emyrk) [reply coderd/users.go]: If the error is fundamental, I wonder if the error should be asserted at the database layer. This is only asserted when keys are returned to the user for visibility. The key was already inserted into the database, and can even be used, since using the token does not trigger this assertion.

**Review by Emyrk — APPROVED** [review]
Will not block on the panic 👍

File: coderd/users.go (hunk 5 above)
Comment (ThomasK33) [reply coderd/users.go]: You're right, we should also assert at insertion. I'll do that.

File: coderd/apikey/apikey.go [hunk 6]
```diff
@@ -102,6 +102,10 @@ func Generate(params CreateParams) (database.InsertAPIKeyParams, string, error)
 		}
 	}
 
+	if len(params.AllowList) == 0 {
+		panic(fmt.Sprintf("developer error: API key %s has empty allow list", keyID))
+	}
```
Comment (Emyrk) [inline coderd/apikey/apikey.go]: I think we should return an error, as this function already supports.

Panics cannot be recovered from. I understand this would mean there is a bug in the usage by the caller (developer). But if this ever triggered in production, it would have no benefit over an error.

Panics should really be a last resort imo.
If an error is returned here, it would still be triggered by any unit test.

File: coderd/users.go (hunk 5 above)
Comment (Emyrk) [reply coderd/users.go]: I agree that if we are to assert with a panic, it should be done earlier. But I still dislike using a panic.

If this panic was to be triggered in production, it would always be better served by an error. An error can be handled and recovered from.

A panic is only beneficial for unit testing. To categorically fail any test that trigger this failed assertion. If a test was to panic, I would hope it would also fail from a returned error.

I still think a panic should only be used if there is 0 possibility to recover from a situation or handle the error more gracefully.

**Review by Emyrk — CHANGES_REQUESTED** [review]
Just the panic stuff left

File: coderd/apikey/apikey.go (hunk 6 above)
Comment (ThomasK33) [reply coderd/apikey/apikey.go]: I understand your concern about using errors instead of panics. You're right that this function already returns errors, and panicking would crash the process.

My reasoning for using panic here is that an empty AllowList signifies a fundamental violation of an invariant. This isn't a case of user input validation or transient failure—it's a fundamental bug in our code that compromises system integrity.
In this scenario, the failure should be immediate and loud, rather than risk a silent issue that might only surface when the token is used.

That said, I understand your point about production resilience. If you feel strongly that returning an error is preferable, I'm open to making that change. The key is that we must prominently alert the developer to their mistake.

Comment (Emyrk) [reply coderd/apikey/apikey.go]: > In this scenario, the failure should be immediate and loud, rather than risk a silent issue that might only surface when the token is used.

An error or a panic will both fail in the same scenario. There should never be a user case where an error would not be escalated, and a panic is required to trigger the failuire log/message.

> rather than risk a silent issue that might only surface when the token is used.

So the error should never be silent.

> My reasoning for using panic here is that an empty AllowList signifies a fundamental violation of an invariant. This isn't a case of user input validation or transient failure—it's a fundamental bug in our code that compromises system integrity.

This is true. But I still think there is no situation where a panic would result in a better outcome than an error.

Comment (ThomasK33) [reply coderd/apikey/apikey.go]: > But I still think there is no situation where a panic would result in a better outcome than an error.

I'm curious about this—if there's supposedly never a situation where panics are appropriate, what are they intended for in Go? They must serve some purpose.

> An error or a panic will both fail in the same scenario. There should never be a user case where an error would not be escalated, and a panic is required to trigger the failuire log/message.

The difference is that the calling code can ignore or mismanage an error, while a panic cannot.

When handling an error, a future developer might:
- Add an unnecessary recovery path
- Log the error and allow execution to continue
- Wrap the error and obscure its root cause

An empty AllowList means we're about to create an API key with an inherent violation of our model and a data inconsistency. This isn't a case of "the database is down" or "the user provided bad input"—it's a situation where our code is fundamentally broken in a way that creates data inconsistencies and introduces undefined behavior in other parts of the system.

When would you use a panic? In your view, what distinguishes an invariant violation from a regular error?

Comment (Emyrk) [reply coderd/apikey/apikey.go]: > I'm curious about this—if there's supposedly never a situation where panics are appropriate, what are they intended for in Go? They must serve some purpose.

There is definitely situations, but they are the edge case.

https://go.dev/doc/effective_go#panic

> If the problem can be masked or worked around, it's always better to let things continue to run rather than taking down the whole program. 

I think it really comes down to, "should this error bring down the executable". In our case, should this crash coderd? 

I don't think so. I think this error would likely occur because some legacy api key slipped through the cracks. In which case, that api key would be rejected, but the application as a whole would not be taken down.

In the case we have too, if we panic, then the app would constantly panic everytime the apikey is used. 

In production we have a `Recover` function at the middleware level to prevent this, because panics should really never occur. So we stop them from affecting the rest of the app (if some library or divide by 0 kinda bug slips through).

------------

There are 2 places we are panicing in this PR.

The first is `convertAPIKey` which is just converting a database type to a codersdk type. In this case, if we panic here I think we doing a larger dis-service then just translating the type.

There is nothing you can do to "fix" the problem here. The database type has an empty allow list, so if a user is requesting to view this database row, why error at all? Why not just return the "corrupted" row as is. Letting the user/admin view the honest APIKey as it exists.

We do not use `codersdk.*` types in our authentication logic (use of the key)

The second panic place is in `Generate`. This is I think just an oversight, because we do 

```
if len(params.AllowList) == 0 {
  params.AllowList = database.AllowList{{Type: policy.WildcardSymbol, ID: policy.WildcardSymbol}}
}

//  ... code ...
if len(params.AllowList) == 0 {
  panic(fmt.Sprintf("developer error: API key %s has empty allow list", keyID))
}
```

So do we allow an empty list or do we not?

Comment (ThomasK33) [reply coderd/apikey/apikey.go]: > I don't think so. I think this error would likely occur because some legacy api key slipped through the cracks. In which case, that api key would be rejected, but the application as a whole would not be taken down.

In this case, we're referring to the return value of the `func Generate(params CreateParams)` function. If this assertion fails, it means we've written incorrect code. This type of error cannot be handled within the code itself because it's not transient or recoverable; the only solution is to upgrade to a newer version where the issue is resolved.

> Why not just return the "corrupted" row as is. Letting the user/admin view the honest APIKey as it exists.

Are you suggesting we introduce undefined behavior into the code?

Returning "corrupted" entries will eventually lead to undefined behavior, as it would violate assumptions the code relies on. This could result in errors in other subsystems and make it much harder to trace the source of issues.

> The second panic place is in Generate. This is I think just an oversight, because we do

No, it isn’t. The explicit assertion ensures that the Generate function never returns an empty allow list. While it’s evident that an empty list won’t be returned in this case, future refactoring could make this less clear. The assert preserves correctness.

---

These assertions do not handle errors. They focus on asserting correctness and reducing undefined behavior in production code.

Comment (Emyrk) [reply coderd/apikey/apikey.go]: > No, it isn’t. The explicit assertion ensures that the Generate function never returns an empty allow list. While it’s evident that an empty list won’t be returned in this case, future refactoring could make this less clear. The assert preserves correctness.

By that same logic though, if someone removes

```golang
	if len(params.AllowList) == 0 {
		params.AllowList = database.AllowList{{Type: policy.WildcardSymbol, ID: policy.WildcardSymbol}}
	}
```

What prevents them from removing the panic as well? If we are trying to protect a function from changing its API usage, an error is still a common way to indicate an api is being misused. In this case calling `Generate` with an empty list.

Golang libraries do no panic on mis-use, they return errors. This kind of error will guarantee `Generate` never returns an empty list just as much as a panic. Errors should not be ignored, and this will be raised to the eventual caller with an error stack.

```golang
	if len(params.AllowList) == 0 {
		return database.InsertAPIKeyParams{}, "", xerrors.Errorf("api keys require a non-empty allow list")
	}
```

---

> Are you suggesting we introduce undefined behavior into the code?

> Returning "corrupted" entries will eventually lead to undefined behavior, as it would violate assumptions the code relies on. This could result in errors in other subsystems and make it much harder to trace the source of issues.

It is not undefined. `convertAPIKey` is used by 3 routes.

```golang
// @Router /users/{user}/keys/{keyid} [get]
// @Router /users/{user}/keys/tokens/{keyname} [get]
// @Router /users/{user}/keys/tokens [get]
```

### Panic

If we panic, these routes will explode if we have an empty allow list. So if we have some key `EmptyKey`, fetching it just always gets `"An internal server error occurred."` with 0 details.

The panic trace is logged. 

The admin has to look through the logs to identify this panic occurs, but they will have not info on the route, the key, etc. The request details are not logged in a panic recover.

^ I do not see this as a good debugging experience for the admin, or a good experience for the user. As there is no actionable info.

### Error

If we error, these routes return a `http.BadRequest` with some message like `api key has empty allow list`.

The client now has information on why an error occurs. These error can be traced by an admin to the `convertAPIKey`. 

^ A better debug experience, but still not great because the error does not have the key name/id.

### No error, just allow `convertAPIKey` to return `[]`

These routes return APIKeys with empty allow lists. These routes are just fetching data from the db and converting it to an HTTP response. 

If an empty allow list will really break our system, a panic or error at this point it too late. We need to panic where we fetch and use the key. Maybe even in the `authorize` call in `/rbac`. 

The empty list has no function at `convertAPIKey`.

**However** I disagree that behavior is undefined at all. An empty allow list has 0 access. That is defined behavior. So allowing an empty allow list around the codebase **is defined**. I agree from a user experience it does not make sense, in which case I think we can raise helpful errors to correct their usage.

Comment (johnstcn) [reply coderd/apikey/apikey.go]: We should not panic here. Let's instead return a sentinel error that clearly states that it's a "Developer error".

Comment (ThomasK33) [reply coderd/apikey/apikey.go]: > By that same logic though, if someone removes

That's not the point, and this is not relevant in this context.

Your statement argues, rephrased, "Why add validation to a form if someone could simply remove that validation?"

In these hypothetical scenarios, inexperienced developers making careless changes could introduce undefined behavior and incorrect code. However, this does not diminish the value of validation and assertions.

> Golang libraries do no panic on mis-use, they return errors. 

This is not a library; it's application code. We must prioritize correctness above all else and work to minimize undefined behavior. The focus isn't on writing a library, but on ensuring our application code is correct.

> However I disagree that behavior is undefined at all. An empty allow list has 0 access. That is defined behavior. So allowing an empty allow list around the codebase is defined. I agree from a user experience it does not make sense, in which case I think we can raise helpful errors to correct their usage.

That is incorrect. We have clearly stated, and based on our assumptions, that empty allow lists are not permitted, as they make neither conceptual nor logical sense—the API key would be rendered useless. Database migrations ensure that any legacy API keys are properly grandfathered in, leaving no empty allow lists post-migration. Furthermore, no HTTP handler or function ever creates an API key with an empty allow list.

You have also made these points yourself in:
- https://app.graphite.dev/github/pr/coder/coder/19964/feat-add-allow_list-to-resource-scoped-API-tokens#comment-PRRC_kwDOGkVX1s6PMEEv
- https://app.graphite.dev/github/pr/coder/coder/19964/feat-add-allow_list-to-resource-scoped-API-tokens#comment-PRRC_kwDOGkVX1s6PMBzr

> We should not panic here. Let's instead return a sentinel error that clearly states that it's a "Developer error".

How would you handle this error in the code? The caller cannot handle this sentinel error if the code is incorrect. So, if an error can't be handled, is it truly an error or an impossible situation instead?

Comment (Emyrk) [reply coderd/apikey/apikey.go]: > That is incorrect. We have clearly stated, and based on our assumptions, that empty allow lists are not permitted, as they make neither conceptual nor logical sense

We have made the code such that empty allow lists do not get put into the data model. But there is no assertion in the db layer or the rbac layer for empty allow lists.

The rego layer & db accept empty allow lists.

> This is not a library; it's application code.

[`Generate`](https://github.com/coder/coder/blob/44744f745f41914bc5ee6538aaca4c9af77c7454/coderd/apikey/apikey.go#L47-L47) is in the `apikey` package. This package is a library for use within the application.

The line between application & library code is pretty thin imo. We should expect packages within our mono repo to be contained and have defined api surfaces.

> How would you handle this error in the code? The caller cannot handle this sentinel error if the code is incorrect. So, if an error can't be handled, is it truly an error or an impossible situation instead?

I stated how we can handle the code for `convertAPIKey`. I say we accept the empty list in the API output.

For `Generate` we already handle the empty allow list and default to `*:*`.

We can add a unit test to make sure the return is never going to be an empty list if we are worried about a future developer removing the default behavior.

> inexperienced developers making careless changes could introduce undefined behavior and incorrect code

This is always true though. We can add a unit test to add code somewhere else to ensure the assertion is not easily removed.

There is 2 pieces of code that handle the `len(params.AllowList) == 0`. 

One updates the default value to `database.AllowList{{Type: policy.WildcardSymbol, ID: policy.WildcardSymbol}}`

The other panics.

How does the panic ensure an inexperienced dev does not make a mistake more than the previous defaulting code? We can add comments to add some edification as to the rational.

Comment (Emyrk) [reply coderd/apikey/apikey.go]: By this same panic logic, should we also panic on `params.ExpiresAt` being in the future and `params.LifetimeSeconds` not being 0? They have the same behavior, where omitting a value is defaulted to something else.

This is a very common pattern for us to assume zero values have some other default. We do not have panic assertions to make sure these values are actually initialized.

We do have unit tests to verify the defaults are applied.

Comment (ThomasK33) [reply coderd/apikey/apikey.go]: > We have made the code such that empty allow lists do not get put into the data model. But there is no assertion in the db layer or the rbac layer for empty allow lists.
> The rego layer & db accept empty allow lists.

I agree this is a shortcoming, and we should add those checks and assertions in future PRs.

> Generate is in the apikey package. This package is a library for use within the application.

This is not a general-purpose library. It is a function and a package that is highly specific application code for coder and usable only within that context, not in other applications. It hasn't been released separately; this internal code has no go.mod and no stability guarantees for external use.

> I say we accept the empty list in the API output.

Why would we allow an impossible state?
If an API key has no allow list entries, what is its purpose, especially without the ability to update it post-creation?

The HTTP handler will reject any request to create such a key during validation of the request.
A key with an empty allow list may never exist.

 > I stated how we can handle the code for convertAPIKey. I say we accept the empty list in the API output.

That does not explain how the application itself handles this incorrect case. How can coderd recover from or resolve this error while operating? How can coderd "fix" or call the `Generate` function differently if it starts returning empty allow lists?

> There is 2 pieces of code that handle the len(params.AllowList) == 0.
> One updates the default value to database.AllowList{{Type: policy.WildcardSymbol, ID: policy.WildcardSymbol}}
> The other panics.

That's not correct. Only one piece of code handles the case where `len(params.AllowList) == 0`, and it inserts the default wildcard entry. 
The other asserts that the allow list always contains at least one entry.

In the `Generate` case, it's evident that the assertion will never trigger. However, once the code is refactored in the future and a developer inadvertently introduces an error, this assertion will serve as a safeguard, flagging those violated assumptions immediately.

> should we also panic on params.ExpiresAt being in the future and params.LifetimeSeconds not being 0

We should be, good point.

>This is a very common pattern for us to assume zero values have some other default. We do not have panic assertions to make sure these values are actually initialized.

We have assertions throughout the project and do not always assume default values.

- https://github.com/coder/coder/blob/HEAD/agent/agentcontainers/subagent.go#L156-L158
- https://github.com/coder/coder/blob/HEAD/cli/exp_mcp.go#L977-L979
- https://github.com/coder/coder/blob/HEAD/codersdk/insights.go#L24-L33
- https://github.com/coder/coder/blob/HEAD/coderd/database/modelmethods.go#L673-L683
- https://github.com/coder/coder/blob/HEAD/coderd/awsidentity/awsidentity.go#L57-L66

Considering the use of asserts and panics outside of tests in our codebase, I would argue that this concept is not unfamiliar to us.

```
rg 'panic\(' --type go \
    --glob '!*_test.go' \
    --glob '!**/testutil/**' \
    --glob '!**/coderdtest/**' \
    --glob '!**/*test*/**' \
    --count-matches | awk -F: '{sum += $2} END {print sum}'
```

File: coderd/provisionerdserver/provisionerdserver_test.go [hunk 7]
```diff
@@ -4031,9 +4031,16 @@ func TestServer_ExpirePrebuildsSessionToken(t *testing.T) {
...
-		user            = dbgen.User(t, db, database.User{})
-		template        = dbgen.Template(t, db, database.Template{
+		srv, db, ps, pd = setup(t, false, &overrides{
+			deploymentValues: &codersdk.DeploymentValues{
+				Sessions: codersdk.SessionLifetime{
+					DefaultTokenDuration: serpent.Duration(24 * time.Hour),
+					MaximumTokenDuration: serpent.Duration(48 * time.Hour),
+				},
+			},
+		})
```
Comment (Emyrk) [inline coderd/provisionerdserver/provisionerdserver_test.go]: We can just do this in `setup`
`deploymentValues := coderdtest.DeploymentValues(t)`

Not have to set overrides

Comment (ThomasK33) [reply coderd/provisionerdserver/provisionerdserver_test.go]: Moved it to another PR: https://github.com/coder/coder/pull/20444

**Review by Emyrk — APPROVED** [review]
